convergent/convergent.go

```diff
@@ -0,0 +1,142 @@
+// Package convergent handles convergent encryption, also known as content hash keying, it is a cryptosystem that produces identical ciphertext
+// from identical plaintext files. This has applications in cloud computing to remove duplicate files from storage without the provider having access to the encryption keys.
+// The combination of deduplication and convergent encryption was described in a backup system patent filed by Stac Electronics in 1995.
+// Convergent encryption is open to a "confirmation of a file attack" in which an attacker can effectively confirm whether a target possesses
+// a certain file by encrypting an unencrypted, or plain-text, version and then simply comparing the output with files possessed by the target.
+
+package convergent
+
+import (
+	"crypto/sha1"
+	"crypto/sha512"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"sort"
+
+	"github.com/ovh/symmecrypt"
+	"github.com/ovh/symmecrypt/stream"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+type ConvergentEncryptionConfig struct {
+	Identifier  string `json:"identifier,omitempty"`
+	Timestamp   int64  `json:"timestamp,omitempty"`
+	Cipher      string `json:"cipher"`
+	LocatorSalt string `json:"localtor_salt"`
+	SecretValue string `json:"secret_value"`
+}
+
+func LoadKeyFromConfig() (symmecrypt.Key, error) {
+	return nil, nil
+}
+
+func NewKey(h hash.Hash, cfgs ...ConvergentEncryptionConfig) (symmecrypt.Key, error) {
```

Comment:
h shouldnt be hash.Hash, it should be a string representation.
A common workflow would be to store this locator+key to be able to re-read the data later, at which point you cannot easily re-instantiate a hash.Hash from an io.Reader via NewHash() (all you have is a string representation).